### PR TITLE
Refactor detection scripts to be more flexible

### PR DESCRIPTION
**Pull request by lewfish**
This PR refactors the detection script to be more flexible and easy to use. 

* Use click to construct CLIs in Python and convert Bash scripts to Python. 
* Allow users to specify paths to all files instead of following a rigid convention. 
* Paths can be for files on S3 or the local file system
* Lists of TIFFs can be provided which are converted to VRTs dynamically
* Multiple training chip sets can be converted into a single TFRecord. 
* Fixes the problem with Batch jobs not being killable via the console.

Connects #126
Connects #127 
Connects #129 
Connects #107

### CODE REVIEW

File: src/detection/rv/commands/predict.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+from rv.commands.settings import planet_channel_order
+from rv.commands.utils import (
+    download_if_needed, upload_if_needed, get_local_path, build_vrt,
+    make_temp_dir)
+
+
+@click.command()
+@click.argument('inference_graph_uri')
+@click.argument('label_map_uri')
+@click.argument('agg_predictions_uri')
```
Comment (notthatbreezy) [inline src/detection/rv/commands/predict.py]: Cool, it will be nice to specify outputs directly

**Comment by lewfish** [comment]
In addition to handling explicit lists of TIFFs to turn into VRTs, should we also support lists of length one that reference a directory, which will indicate that all the TIFFs in that directory should be downloaded?

Also, if we want to generate training chips for a set of sets of TIFFs (ie. multiple VRTs), we need to specify a list of lists on the command line which doesn't seem supported by click and would be unwieldy. In this case, can we assume that each set of tiffs is in a directory, so that we just need to specify a list of directories?

File: src/detection/scripts/batch_submit.py [hunk 2]
```diff
@@ -1,14 +1,23 @@
...
 s3_bucket = environ.get('S3_BUCKET')
 
 
-def submit(branch_name, command_args, attempts=3, cpu=False):
+@click.command()
+@click.argument('branch_name', help='Branch with code to run on Batch')
+@click.argument('command', help='Command in quotes to run on Batch')
+@click.option('--attempts', default=3, help='Number of times to retry job')
+@click.option('--cpu', is_flag=True, help='Use CPU EC2 instances')
+def batch_submit(branch_name, command_args, attempts=3, cpu=False):
```
Comment (notthatbreezy) [inline src/detection/scripts/batch_submit.py]: Is this still needed? Is this what gets run on the batch instance then?

File: src/detection/README.md [hunk 3]
```diff
@@ -13,20 +13,20 @@ In order to train a model, you must convert a set of GeoTIFF and GeoJSON file wi
 to use a default ordering of `2 1 0` for Planet imagery, or can be specified for different orderings.
 
 ```
-python scripts/tiff_chipper.py \
-    --input-dir /opt/data/datasets/detection/singapore_ships/train/ \
-    --output-dir /opt/data/datasets/detection/singapore_ships_chips_neg \
-    --chip-size 300 --num-neg-chips 50 --max-attempts 500 --channel-order 2 1 0
+python -m rv.run make_train_chips \
```
[the remark this answers is not in the document]
Comment (lewfish) [reply src/detection/README.md]: I'm going to hold off on setting up the script with setup.py until we're really ready to make things pip installable.

File: src/detection/scripts/batch_submit.py (hunk 2 above)
Comment (lewfish) [reply src/detection/scripts/batch_submit.py]: batch_submit is just the script that creates the Batch job to run the scripts remotely. I think it's still needed.

**Comment by notthatbreezy** [comment]
> In addition to handling explicit lists of TIFFs to turn into VRTs, should we also support lists of length one that reference a directory, which will indicate that all the TIFFs in that directory should be downloaded?

Is this going to be tricky -- will it always be straightforward enough to check it there is an ending slash to know that it is a directory vs file? For what it's work, I think for a long time we'll only be passing in full filepaths.

> Also, if we want to generate training chips for a set of sets of TIFFs (ie. multiple VRTs), we need to specify a list of lists on the command line which doesn't seem supported by click and would be unwieldy. In this case, can we assume that each set of tiffs is in a directory, so that we just need to specify a list of directories?

That's a good point. Does it make sense to be able to arbitrarily create training chips and dump them somewhere with the required CSV/metadata? What I'm getting at is that I could see it being the case that we have a set of training data, generate chips for that. Then we run it against new sets of data -- now we want to generate chips from the validated data as well -- so generating/adding new training data over time in an ad hoc manner seems attractive -- then we wouldn't necessarily have to worry about generating all the training data we need in a single pass.

File: src/detection/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (notthatbreezy) [reply src/detection/README.md]: That makes sense

File: src/detection/scripts/batch_submit.py (hunk 2 above)
Comment (notthatbreezy) [reply src/detection/scripts/batch_submit.py]: Got it - it wasn't clear to me at first glance if this is what kicks off the batch job or if this is supposed to be the command that is running on batch.

**Comment by lewfish** [comment]
> Is this going to be tricky -- will it always be straightforward enough to check it there is an ending slash to know that it is a directory vs file? For what it's work, I think for a long time we'll only be passing in full filepaths.

It would be easy enough if we could assume directories always end in a slash, but if not, we'll have to check that the URI is a directory or file on S3, which complicates things. I'll just leave it as is for now if it's not a priority.

**Comment by lewfish** [comment]
I just need to do a final test on Batch and this should be good to go, from my perspective.

File: src/detection/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (notthatbreezy) [reply src/detection/README.md]: When I run the sample command -- I was trying to create a new set of predictions in a different location, but the job fails `rv.commands.utils.NotFoundException: /opt/data/agg_predictionsasdf.json does not exist`

I don't think the target prediction location should have to exist.

[the remark this answers is not in the document]
Comment (lewfish) [reply src/detection/README.md]: Which command did you run exactly?

[the remark this answers is not in the document]
Comment (notthatbreezy) [reply src/detection/README.md]:
```
 python -m rv.run predict \
    --agg-predictions-debug-uri s3://raster-vision/results/detection/predict/lhf_ships3_crop/agg_predictions.png \
    s3://raster-vision/results/detection/train/lhf_ships_neg0/train/inference_graph.pb \
    s3://raster-vision/datasets/detection/ships_label_map.pbtxt \
    s3://raster-vision/results/detection/predict/lhf_ships3_crop/crop1.tif \
    s3://raster-vision/results/detection/predict/lhf_ships3_crop/crop2.tif \
    /opt/data/agg_predictionsasdf.json
```

[the remark this answers is not in the document]
Comment (lewfish) [reply src/detection/README.md]: I can't replicate this when running in the Docker container. This line should avoid the issue you're describing:
https://github.com/azavea/raster-vision/blob/004abf50c2cc0cc2bc75ff5b78dab00421bef860/src/detection/rv/commands/aggregate_predictions.py#L281

**Pull request merged**